[integration] add admin/developer docs about APIs (#5540)
DIRACGrid/DIRAC: docs/source/DeveloperGuide/APIs/index.rst

.. _apis:

Systems APIs
============

Currently, client-server interaction with DIRAC systems is provided by the HTTP services that writen on the :py:class:`~DIRAC.Core.Tornado.Server.TornadoService` framework.
This framework describe a fixed interface via HTTP POST requests as PRC calls, see :ref:`httpsTornado`.

> Starting with 8.0, an additional :py:class:`~ DIRAC.Core.Tornado.Server.TornadoREST` framework was created, which expands the possibilities for writing access interfaces to DIRAC systems.

Reviewer: ```suggestion
Starting with 8.0, an additional :py:class:`~ DIRAC.Core.Tornado.Server.TornadoREST` module was created, which expands the possibilities for writing access interfaces to DIRAC systems by implementing a REST interface.
```

> Consider the structure:
> -----------------------
> 
> The main change introduced here is the separation in a :py:class:`~ DIRAC.Core.Tornado.Server.private.BaseRequestHandler` class of the basic logic for authorizing the request, search and run the target method in the thread.
> Thus TornadoService and TornadoREST following BaseRequestHandler can describe RPC and REST interfaces respectively::

Reviewer: ```suggestion
``TornadoService`` and ``TornadoREST `` inherit from :py:class:`~ DIRAC.Core.Tornado.Server.private.BaseRequestHandler` class the basic logic for authorizing the request, search and run the target method in the thread::
```

    BaseRequestHandler
    |
    |\_TornadoService
    |
     \_TornadoREST

> From the scheme it is intuitively clear that on the basis of BaseRequestHandler it is possible to create other framework if flexibility of existing is not enough.

Reviewer: ```suggestion
From the scheme it is intuitively clear that on the basis of BaseRequestHandler it is possible to implement other frameworks.
```

New components:
+++++++++++++++

  - :py:class:`~ DIRAC.Core.Tornado.Server.private.BaseRequestHandler` contains the basic logic described here :ref:`httpsTornado` except for the access interface.

>   - :py:class:`~ DIRAC.Core.Tornado.Server.TornadoREST` framework for writing REST API for DIRAC systems

Reviewer: no need.

Author: removed "New components" block

What is the purpose of this?
----------------------------

The purpose of this component is to build DIRAC system status access interfaces based on HTTP requests outside the fixed RPC request interface described in TornadoServices.

The main reason for the implementation of this feature is the need to implementation of API for DIRAC Authorization Server based on OAuth2 framework,
which in turn is an integral part of the implementation of OAuth2 DIRAC authorization.
So currently, there is only one API implementation for the Framework system, which is :py:class:`~ DIRAC.FrameworkSystem.API.AuthHandler`.
For the same reason, the ability to authorize with an access token (see https://datatracker.ietf.org/doc/html/rfc6750#section-2) was added to the authorization steps.

How to write APIs
-----------------

The following is an example of a simple API:

> .. literalinclude::

Reviewer: If you use `.. literalinclude::`  magic you provide a py file: https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-literalinclude

Author: fixed

    from DIRAC.Core.Tornado.Server.TornadoREST import TornadoREST
    from DIRAC.ConfigurationSystem.Client.Helpers import Registry
    class yourEndpointHandler(TornadoREST):

      SYSTEM = "Configuration"
      LOCATION = "/registry"

      @classmethod
      def initializeHandler(cls, infosDict):
        ''' Called only once when the first request for this handler arrives
        '''
        pass

      def initializeRequest(self):
        ''' Called at the beginning of each request
        '''
        pass

      # Specify the default permission for the method
      auth_users = ['authenticated']

      def web_users(self, count: int = 0):
        ''' Your method. It will be available for queries such as https://domain/registry/users?count=1

            TornadoREST will try to convert the received argument `count` to int.

        '''

>         return Registry.getAllUsers()[:count]

Reviewer: run `black` on this code.

Author: done

> .. note:: For the better look of your requests you can listen default https port (443), balancer like a Nginx can help you with this.

Reviewer: "for the better look" ?

Author: rewrited to `If you need to implement the interface on a standard 443 https port, you will need to use a balancer, such as nginx.`

The example described is likely to be sufficient for most writing cases. But here are some additional features, see :py:class:`~ DIRAC.Core.Tornado.Server.private.BaseRequestHandler`:

  - ``USE_AUTHZ_GRANTS`` set the list and order of steps to authorize the request. For example, set ``USE_AUTHZ_GRANTS = ["JWT"]`` to allow access to your endpoint only with a valid access token.

>   - ``AUTH_PROPS`` set the authorization requirements. For example, ``AUTH_PROPS = ['authenticated']`` will allow access only to authenticated users.

Reviewer: are these global variables?

Author: No, its a class variables.

  - in addition to standard S_OK/S_ERROR you can return text, whether the dictionary for example or nothing, the result will be sent with a 200 status.
  - ``path_<my_method>`` will allow you to consider the path parameters as positional arguments of the target method. For example:

    .. code-block:: python

        # It's allow make request like a GET /user/contacts/Bob

>         path_user = ['[A-z]+', '[A-z]+']

Reviewer: do you mean 
```suggestion
        path_user = ["([A-z%0-9-_]*)"]
```
?

Author: yes, thanks. I fixed it to correct regex.

        def web_user(self, option:str name:str):

>           return Registry.getUserOption(name, option)

Reviewer: use 4 spaces

Author: done

  - If your API is complex enough and may include, for example, redirection or additional headers, you can use :py:class:`~ DIRAC.Core.Tornado.Server.private.BaseRequestHandler.TornadoResponse`
    to add all these necessary things, which is thread-save because TornadoResponse will call your actions outside the thread in which this method is executed:

    .. code-block:: python

        from DIRAC.Core.Tornado.Server.private.BaseRequestHandler import TornadoResponse

        # Describe the class as in the example.

          def web_myMethod(self, my_option):
            # Do some thing
            response = TornadoResponse(data)
            response.set_status(201)
            return response

> Since the DIRAC 8.0 is already written in python3, framework takes into account the target method annotation when preparing arguments using ``inspect``, see https://docs.python.org/3/library/inspect.html#inspect.signature.

Reviewer: ```suggestion
The framework takes into account the target method annotations when preparing arguments using the ``inspect`` module, see https://docs.python.org/3/library/inspect.html#inspect.signature.
```

.. _apis:

Systems APIs
=============

Currently, client-server interaction with DIRAC systems is provided by the HTTP services that are coded using the :py:class:`~DIRAC.Core.Tornado.Server.TornadoService` module.
This module implements a fixed interface via HTTP POST requests meant to implement RPC calls, see :ref:`httpsTornado`.
Starting with 8.0, an additional :py:class:`~ DIRAC.Core.Tornado.Server.TornadoREST` module was created, which expands the possibilities for writing access interfaces to DIRAC systems by implementing a REST interface.

> ``TornadoService`` and ``TornadoREST `` inherit from :py:class:`~ DIRAC.Core.Tornado.Server.private.BaseRequestHandler` class the basic logic for authorizing the request, search and run the target method in the thread::

Reviewer: /home/docs/checkouts/readthedocs.org/user_builds/dirac/checkouts/5540/docs/source/DeveloperGuide/APIs/index.rst:10: WARNING: Inline literal start-string without end-string.

Author: fixed

  - ``USE_AUTHZ_GRANTS`` set the list and order of steps to authorize the request. For example, set ``USE_AUTHZ_GRANTS = ["JWT"]`` to allow access to your endpoint only with a valid access token.
  - ``AUTH_PROPS`` set the authorization requirements. For example, ``AUTH_PROPS = ['authenticated']`` will allow access only to authenticated users.
  - in addition to standard S_OK/S_ERROR you can return text, whether the dictionary for example or nothing, the result will be sent with a 200 status.
  - ``path_<my_method>`` will allow you to consider the path parameters as positional arguments of the target method. For example:

    .. code-block:: python

        # It's allow make request like a GET /user/contacts/Bob
        path_user = ['[A-z]+', '[A-z]+']
        def web_user(self, option:str name:str):
            return Registry.getUserOption(name, option)

  - If your API is complex enough and may include, for example, redirection or additional headers, you can use :py:class:`~ DIRAC.Core.Tornado.Server.private.BaseRequestHandler.TornadoResponse`

>     to add all these necessary things, which is thread-save because TornadoResponse will call your actions outside the thread in which this method is executed:

Reviewer: ```suggestion
    to add all these necessary things, which is thread-safe because TornadoResponse will call your actions outside the thread in which this method is executed:
```

    .. code-block:: python

        from DIRAC.Core.Tornado.Server.private.BaseRequestHandler import TornadoResponse

        # Describe the class as in the example.

          def web_myMethod(self, my_option):
            # Do some thing
            response = TornadoResponse(data)
            response.set_status(201)

>             return response

Reviewer: please `black` it

Author: blacked
